ioctl/test_ioctl.c

```diff
@@ -0,0 +1,266 @@
+		}
+		msgRespond(port, &msg, rid);
+	}
+	return retval;
+}
+
+
+static void init(void)
+{
+	char devpath[5] = "test";
+	oid_t dev;
+
+	TEST_ASSERT_EQUAL_INT(EOK, portCreate(&port));
+
+	dev.port = port;
+	dev.id = 0;
+
+	TEST_ASSERT_EQUAL_INT(EOK, create_dev(&dev, devpath));
+
+	return;
+}
+
+
+TEST_GROUP(ioctl);
+
+
+TEST_SETUP(ioctl)
+{
+}
+
+TEST_TEAR_DOWN(ioctl)
+{
+}
+
+
+TEST(ioctl, no_data)
+{
+	int flag;
+	/* Send data to driver by value */
+	while ((ret = ioctl(dd, TIOCSIG, NULL)) < 0) {
+		TEST_ASSERT_EQUAL_INT(-1, ret);
+		usleep(10000);
+	}
+	TEST_ASSERT_EQUAL_INT(0, ret);
+	lseek(fd, 0, SEEK_SET);
+	read(fd, &flag, 4);
+	TEST_ASSERT_EQUAL_INT32(1, flag);
+}
+
+
+TEST(ioctl, in_val)
+{
+	int32_t data = 14;
+	int32_t rdata;
+
+	/* Send data to driver by value */
+	while ((ret = ioctl(dd, TIOCVAL, data)) < 0) {
+		TEST_ASSERT_EQUAL_INT(-1, ret);
+		usleep(10000);
+	}
+	TEST_ASSERT_EQUAL_INT(0, ret);
+	lseek(fd, 0, SEEK_SET);
+	read(fd, &rdata, 4);
+	TEST_ASSERT_EQUAL_INT32(14, rdata);
+}
+
+
+TEST(ioctl, in)
+{
+	int32_t data_in = 20;
+	int32_t rdata;
+
+	/* Send data to driver by pointer */
+	while ((ret = ioctl(dd, TIOCPTR, &data_in)) < 0) {
+		TEST_ASSERT_EQUAL_INT(-1, ret);
+		usleep(10000);
+	}
+	TEST_ASSERT_EQUAL_INT(0, ret);
+	lseek(fd, 0, SEEK_SET);
+	read(fd, &rdata, 4);
+	TEST_ASSERT_EQUAL_INT32(20, rdata);
+}
+
+
+TEST(ioctl, in_64)
+{
+	int8_t data_in[64];
+	memset(data_in, 3, 64);
+	int8_t rdata[64];
+
+	/* Send data to driver by pointer , data big enough to not be copied by ioctl_pack directly into message */
+	while ((ret = ioctl(dd, TIOCPTR2, data_in)) < 0) {
+		TEST_ASSERT_EQUAL_INT(-1, ret);
+		usleep(10000);
+	}
+	TEST_ASSERT_EQUAL_INT(0, ret);
+	lseek(fd, 0, SEEK_SET);
+	read(fd, rdata, 64);
+	TEST_ASSERT_EQUAL_MEMORY(data_in, rdata, 64);
+}
+
+
+TEST(ioctl, out)
+{
+	int32_t data_out;
+
+	/* Get data from driver */
+	while ((ret = ioctl(dd, TIOCOUT, &data_out)) < 0) {
+		TEST_ASSERT_EQUAL_INT(-1, ret);
+		usleep(10000);
+	}
+	TEST_ASSERT_EQUAL_INT32(15, data_out);
+}
+
+
+TEST(ioctl, out_64)
+{
+	int8_t data_expected[64];
+	memset(data_expected, 5, 64);
+	int8_t data_out[64];
+
+	/* Get data from driver */
+	while ((ret = ioctl(dd, TIOCOUT2, data_out)) < 0) {
+		TEST_ASSERT_EQUAL_INT(-1, ret);
+		usleep(10000);
+	}
+	TEST_ASSERT_EQUAL_MEMORY(data_expected, data_out, 64);
+}
+
+
+TEST(ioctl, no_device)
+{
+	errno = 0;
+	int ret = ioctl(123, 0, NULL);
+	TEST_ASSERT_EQUAL_INT(-1, ret);
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+}
+
+
+TEST_GROUP_RUNNER(ioctl)
+{
+	fd = open(FILENAME, O_RDWR | O_CREAT, 0666);
+	/* other thread - create device server */
+	init();
+	pthread_create(&tid, NULL, thread, NULL);
+
+	while ((dd = open("/dev/test", O_RDWR)) < 0) {
+		usleep(10000);
+	}
+
+	/* main thread - run test cases */
+	RUN_TEST_CASE(ioctl, no_device);
+	RUN_TEST_CASE(ioctl, no_data);
+	RUN_TEST_CASE(ioctl, in_val);
+	RUN_TEST_CASE(ioctl, in);
+	RUN_TEST_CASE(ioctl, in_64);
```

Comment:
I know that it's hard sometimes, but try to create more suitable names, it would be perfect to bind it with incoming requirements from the documentation.